chain/consensus/mir/request_cache.go

```diff
@@ -0,0 +1,62 @@
+package mir
+
+import (
+	"sync"
+)
+
+// Request cache is the simplest temporal cache to store mapping between request hashes and client requests.
+func newRequestCache() *requestCache {
+	return &requestCache{
+		cache:          make(map[string]ClientRequest),
+		handledClients: make(map[string]bool),
+	}
+}
+
+type ClientRequest struct {
+	Request  Request
+	ClientID string
+}
+
+type requestCache struct {
+	lk sync.Mutex
+
+	cache          map[string]ClientRequest
+	handledClients map[string]bool
+}
+
+// addIfNotExist adds the request if key h doesn't exist .
+func (c *requestCache) addIfNotExist(clientID, h string, r Request) (exist bool) {
+	c.lk.Lock()
+	defer c.lk.Unlock()
+
+	if exist = c.handledClients[clientID]; !exist {
+		c.cache[h] = ClientRequest{r, clientID}
+		c.handledClients[clientID] = true
+	}
+	return
+}
+
+// getDel gets the target request by the key k and deletes the keys.
+func (c *requestCache) getDel(h string) (Request, bool) {
```

Comment:
In general, does the cache only support storing one request per client at a time?

Reply:
Yes, it does. All messages are stored in mempool, we don't add a message from mempool for the client if the cache already has a message for that client.